Docs: Update dev environment setup (#24871)
WordPress/gutenberg: docs/designers-developers/developers/tutorials/devenv/readme.md

# Development Environment

This guide is for setting up your local environment for JavaScript development for creating plugins and tools to extend WordPress and the block editor.

A development environment is a catch-all term for what you need setup on your computer to work. The three main pieces needed for our development environment are:

1. Node/NPM Development Tools
2. WordPress Development Site
3. Code Editor

## Quickstart

Here is a summary of the guide. See each section for additional details and explanations.

**1. Install Node development tools**

Download and install [Node Version Manager](https://github.com/nvm-sh/nvm) (nvm)

```
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
```

Quit and restart terminal
```
nvm install --lts
```

**2. WordPress Development Site**

First download, install, and start [Docker Desktop](https://www.docker.com/products/docker-desktop) following the instructions for your OS.

- Install WordPress environment tool

```
npm -g install @wordpress/env
```

Start the environment:

 ```

>  wp-env start

Reviewer: Should we clarify that this command would need to be run by `cd`ing to a different directory if they wish to use it for an existing plugin/codebase? (e.g. for developing gutenberg?)

The instance will be different for each time they run it, so any changes they make to this instance would not apply if they started a different one from the Gutenberg directory.

Reviewer: if the user just does this first without realizing any of the later information, they might get conflicting port numbers, which is why I mention it.

Author: I can try to clarify, we can also add a troubleshooting section, list a couple of common errors

 ```

You will have a full WordPress site installed, navigate to: http://localhost:8888/ the defualt user is "admin" and default password is "password"

> 

Reviewer 2: As the above comment is resolved I am adding another comment.

Ahhh....

` wp-env start`

You will have a full WordPress site installed, navigate to: http://localhost:8888/ the defualt user is "admin" and default password is "password"

----> Additional information which is good to know about. 

Another way to open the local WordPress site after it has been started up is through Docker. 
Open Docker and go to the Dashboard. 

![Docker-Dashboard](https://user-images.githubusercontent.com/5323259/92227783-13f51780-eea7-11ea-97fe-7730bf166fe0.png)

Click where it says RUNNING. A new screen opens and it will show the containers (what is a container?). 

![Docker containers](https://user-images.githubusercontent.com/5323259/92227672-e4dea600-eea6-11ea-9a9a-779c0ccfc3b6.png) 

Hovering the cursor over one of the containers shows various options. Among them "OPEN IN BROWSER"
Clicking it will then open the local site in the port mentioned.

Author: I think adding the Docker information is confusing

Reviewer 2: What about saying something like this:
"Another option would be to open the browser from Docker by entering the Dashboard, then clicking the stack and selecting a container to open."  Or something like that.

My purpose in adding it is to share information that one can use Docker for various purposes, and this is one of them. Perhaps we can add more Docker information in another place in the docs?

Reviewer 2: We could add it to the more details instructions area?

> 

Reviewer 2: the defualt user = the default user.

Author: Fixed.

**3. Code Editor**

You can use any code editor, [Visual Studio Code](https://code.visualstudio.com/) is a popular open-source tool, follow instructions on their site for your OS.

> 

Reviewer 2: See also the Code Editor section below for additional information. (As we here need to let the user know that there is another Code Editor section below. I am wondering if these should be merged into one Code Editor section. It feels a bit strange having two.)

> You can use any code editor, [Visual Studio Code](https://code.visualstudio.com/) is a popular open-source tool, follow instructions on their site for your OS.

Reviewer: ```suggestion
You can use any text editor to write code. For example, Visual Studio Code](https://code.visualstudio.com/) is a popular open-source editor. You can follow instructions on their site to install it for your OS.
```

# Development Environment

This guide is for setting up your local environment for JavaScript development for creating plugins and tools to extend WordPress and the block editor.

> 

Reviewer 2: This guide is for setting up a local JavaScript development environment for creating plugins and tools to extend WordPress and the block editor. We can also use this setup to view and create Pull Requests (PRs) associated with the block editor.

Author: I don't think we need to add the GitHUB PR process here. What they do with the environment is up to them.

A development environment is a catch-all term for what you need setup on your computer to work. The three main pieces needed for our development environment are:

1. Node/NPM Development Tools
2. WordPress Development Site
3. Code Editor

## Quickstart

For experienced developers, here is a summary of the guide, see each section for more details and explanations.

> 

Reviewer 2: (Skip the for experienced developers.... as the quickstart is for anyone wanting to quickly get into setting up a local dev environment site.)

My suggestion:
"Here is a summary of the guide. See each section for additional details and explanations."

Author: Changed :+1:

**1. Install Node development tools**

Download and install [Node Version Manager](https://github.com/nvm-sh/nvm) (nvm)

```
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
```

> 

Reviewer 2: Result:

```
paaljoachimromdahl@Paals-MacBook-Pro ~ % curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13527  100 13527    0     0   3714      0  0:00:03  0:00:03 --:--:--  3714
=> Downloading nvm from git to '/Users/paaljoachimromdahl/.nvm'
=> Cloning into '/Users/paaljoachimromdahl/.nvm'...
remote: Enumerating objects: 290, done.
remote: Counting objects: 100% (290/290), done.
remote: Compressing objects: 100% (257/257), done.
remote: Total 290 (delta 35), reused 97 (delta 20), pack-reused 0
Receiving objects: 100% (290/290), 163.27 KiB | 781.00 KiB/s, done.
Resolving deltas: 100% (35/35), done.
=> Compressing and cleaning up git repository

=> nvm source string already in /Users/paaljoachimromdahl/.zshrc
=> bash_completion source string already in /Users/paaljoachimromdahl/.zshrc
=> Close and reopen your terminal to start using nvm or run the following to use it now:

export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
paaljoachimromdahl@Paals-MacBook-Pro ~ % 
```

Author: Looks good

Quit and restart terminal
```
nvm install --lts
```

> 

Reviewer 2: Result:

```
paaljoachimromdahl@Paals-MacBook-Pro ~ % nvm install --lts
Installing latest LTS version.
Downloading and installing node v12.18.3...
Downloading https://nodejs.org/dist/v12.18.3/node-v12.18.3-darwin-x64.tar.gz...
######################################################################### 100.0%
Computing checksum with shasum -a 256
Checksums matched!
Now using node v12.18.3 (npm v6.14.6)
Creating default alias: default -> lts/* (-> v12.18.3)
paaljoachimromdahl@Paals-MacBook-Pro ~ % 
```

Author: Looks good

**2. WordPress Development Site**

First download, install, and start [Docker Desktop](https://www.docker.com/products/docker-desktop) follows instructions on their site for your OS.

> 

Reviewer 2: follows = and follow instructions for your OS. 

As one downloads and installs Docker. After having installed Docker one starts it up and there is a tutorial that also begins. One can choose to skip the tutorial.

Author: :+1:

> 

Reviewer 2: follows = and follow instructions for your OS.

There are some instructions to follow after starting up Docker for the first time. As one inserts the user password and a tutorial shows up which one can skip at any time.

It would be good with some info on why one is using Docker. What is it used for?

Author: This is in the quick start, I added a description below.

- Install WordPress environment tool

```
npm -g install @wordpress/env
```

> 

Reviewer 2: Is it good with some info on what wordpress/env is.

Reviewer 2: Result:

```
paaljoachimromdahl@Paals-MacBook-Pro ~ % npm -g install @wordpress/env
npm WARN deprecated request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
/Users/paaljoachimromdahl/.nvm/versions/node/v12.18.3/bin/wp-env -> /Users/paaljoachimromdahl/.nvm/versions/node/v12.18.3/lib/node_modules/@wordpress/env/bin/wp-env

> nodegit@0.26.5 install /Users/paaljoachimromdahl/.nvm/versions/node/v12.18.3/lib/node_modules/@wordpress/env/node_modules/nodegit
> node lifecycleScripts/preinstall && node lifecycleScripts/install

[nodegit] Running pre-install script
[nodegit] Running install script
node-pre-gyp
WARN Using request for node-pre-gyp https download
[nodegit] Success: "/Users/paaljoachimromdahl/.nvm/versions/node/v12.18.3/lib/node_modules/@wordpress/env/node_modules/nodegit/build/Release/nodegit.node" is installed via remote
[nodegit] Completed installation successfully.

> nodegit@0.26.5 postinstall /Users/paaljoachimromdahl/.nvm/versions/node/v12.18.3/lib/node_modules/@wordpress/env/node_modules/nodegit
> node lifecycleScripts/postinstall

+ @wordpress/env@1.6.0
added 274 packages from 236 contributors in 82.855s
paaljoachimromdahl@Paals-MacBook-Pro ~ % 

```

Author: This is in the quick start, if they want details they can read below that explains.

Open a text editor and create a `.wp-env.json` file with the following content:

> 

Reviewer 2: Open a text editor and create a `.wp-env.json` file. 
Suggest a place to save the file.

Add the following content:

Reviewer 2: Add the following content:

Author: There is an expectation that they know how to create and view files on their computer.

```json
{
	"core": null
}

> ```

Reviewer 2: Btw the below comment showing my run through shows my tests saving the json file.

Remove the word json from the above code. So it becomes:

```
{
	"core": null
}
```

As the word json in the file gave this error:
```
paaljoachimromdahl@Paals-MacBook-Pro ~ % wp-env start
✖ Invalid .wp-env.json: Unexpected token j in JSON at position 0
```

Author: I'll be removing this step once confirmed.

> 

Reviewer 2: Creating the json file:
![Create-wp-env-json-file](https://user-images.githubusercontent.com/5323259/91844429-bdde6500-ec57-11ea-99c4-063564a58580.png)

(It is not wise to use the regular text editor on the Mac as it automatically inserts an extension of .rtf which I later on be removed.) 

To show hidden files on the Mac: Press Command (cmd) + Shift + . (dot) to show or hide hidden files.

Reviewer 2: Result:

```
paaljoachimromdahl@Paals-MacBook-Pro ~ % wp-env start
✖ Invalid .wp-env.json: Unexpected token \ in JSON at position 1
paaljoachimromdahl@Paals-MacBook-Pro ~ % 
```

Trying the troubleshooting instructions further below:

If you see the error: `No .wp-env.json file found` or `Invalid .wp-env.json`
- You need to create a valid .wp-env.json file. Try a file with just `{}`

I also quit and restarted terminal.
The error remains. I will use another editor to create the file. Looking at the recommended list, and I am downloading Atom.

Author: I will test and probably can remove the .wp-env.json steps.

This is a minimal file, you can also specify various features and plugins, see documentation for full details. The `"core": null` parameter indicates no version specified, so it defaults to the latest version.

> 

Reviewer 2: See documentation for full details. 
I am guessing you mean see the detailed walk through for full details.

> 

Reviewer 2: Which documentation? Should documentation be a link?

Author: Added wp-env and link

 Start the environment:

 ```
 wp-env start
 ```

> 

Reviewer 2: Result:
```
paaljoachimromdahl@Paals-MacBook-Pro ~ % wp-env start
✔ WordPress started. (in 27s 456ms)
```

Author: :+1:

You will have a full WordPress site installed, navigate to: http://localhost:8888/ the defualt user is "admin" and default password is "password"

> 

Reviewer 2: In which folder is the localhost web site located?

Author: It is in Docker, not a folder.

Reviewer 2: Ahhh....

` wp-env start`

You will have a full WordPress site installed, navigate to: http://localhost:8888/ the defualt user is "admin" and default password is "password"

Another way to open the local WordPress site after it has been started up is through Docker. 
Open Docker and go to the Dashboard. 

![Docker-Dashboard](https://user-images.githubusercontent.com/5323259/92227783-13f51780-eea7-11ea-97fe-7730bf166fe0.png)

Click where it says RUNNING. A new screen opens and it will show the containers (what is a container?). 

![Docker containers](https://user-images.githubusercontent.com/5323259/92227672-e4dea600-eea6-11ea-9a9a-779c0ccfc3b6.png) 

Hovering the cursor over one of the containers shows various options. Among them "OPEN IN BROWSER"
Clicking it will then open the local site in the port mentioned.

The WordPress [wp-env package](https://www.npmjs.com/package/@wordpress/env) lets you set up a local WordPress environment for building and testing plugins and themes, without any additional configuration.

The `wp-env` tool uses Docker to create a virtual machine to that runs the WordPress site. There are instructions available for installing Docker on [Windows 10 Pro](https://docs.docker.com/docker-for-windows/install/), [all other versions of Windows](https://docs.docker.com/toolbox/toolbox_install_windows/), [macOS](https://docs.docker.com/docker-for-mac/install/), and [Linux](https://docs.docker.com/v17.12/install/linux/docker-ce/ubuntu/#install-using-the-convenience-script).

> After you have installed Docker, go ahead and install the `wp-env` tool. This command will install the tool globally, this means you can run it from any directory:

Reviewer: ```suggestion
After you have installed Docker, go ahead and install the `wp-env` tool. This command will install the tool globally, which means you can run it from any directory:
```

> 1.6.0
```

> The `wp-env` script is used to create a Docker WordPress environment. You can use this script to start an environment from your plugin directory, for example if you are following the create block tutorial, this would be in the generated directory like so:

Reviewer: ```suggestion
The `wp-env` script is used to create a Docker WordPress environment. You can use this script to start an environment with your plugin activated by running it from the directory containing your plugin. For example if you are following the create block tutorial, this would be in the generated directory like so:
```

```sh
npx @wordpress/create-block starter-block
cd starter-block
wp-env start
```

You can access your environment in your browser at: [http://localhost:8888/](http://localhost:8888/), the default username is `admin` and default password is `password`. For more information controlling the Docker environment see the [@wordpress/env package readme](/packages/env/README.md).

> 

Reviewer 2: Let's add some Docker info here.
Another option would be to open the browser through Docker. Go to the Docker Dashboard, then click the stack and select a container to open.
---- or something like that. Then the following information can be added after it.... "For more information about controlling the Docker environment see the...."

Full text as I see it:
You can access your environment in your browser at: [http://localhost:8888/](http://localhost:8888/), the default username is `admin` and default password is `password`. Another option would be to open the browser through Docker. Go to the Docker Dashboard, then click the stack and select a container to open.
For more information about controlling the Docker environment see the [@wordpress/env package readme](/packages/env/README.md).

Author: I don't think we need to add Docker info. It is confusing and gives them no additional benefit than just typing that address in their browser. Also, the directions only work if you're running Docker Desktop, which some like myself do not have that interface, I just run the daemon service.

Reviewer 2: Ahhh gotcha! I am letting it go..:)

> When using the script while developing a single plugin, when you run `wp-env start` it will mount and activate the plugin automatically. Note: This also works for themes, run from the directory you are developing the theme.

Reviewer: ```suggestion
When using the script while developing a single plugin, `wp-env start` can mount and activate the plugin automatically when run from the directory containing the plugin. Note: This also works for themes when run from the directory in which you are developing the theme.
```

If you run `wp-env start` from a directory that is not a plugin or theme, a generic WordPress environment will be created. The script will display the following warning, it is fine if this is your intention.

```
!! Warning: could not find a .wp-env.json configuration file and could not determine if 'DIR' is a WordPress installation, a plugin, or a theme.
```

> The `.wp-env.json` file is a configuration file used to create an environment that works with multiple plugins and/or themes. See the [@wordpress/env package for additional details](/packages/env/README.md#wp-envjson).

Reviewer: ```suggestion
You can use the `.wp-env.json` configuration file to create an environment that works with multiple plugins and/or themes. See the [@wordpress/env package for additional details](/packages/env/README.md#wp-envjson).
```

#### Troubleshooting

> If you run into an issue, when trying to run `wp-env` a common one is `Error while running docker-compose command.`

Reviewer: ```suggestion
A common issue when running `wp-env` is `Error while running docker-compose command.`
```

- Check that Docker Desktop is started and running.
- Check Docker Desktop dashboard for logs, restart, or remove existing VMs.

If you see the error: `Host is already in use by another container`

- The container is already running, or another one is. You can stop an existing container running use `wp-env stop` from the directory you started it.

### Alternative to Docker

> Docker is just one method to run a local WordPress environment. Block development and extending WordPress is done using normal plugins, so any WordPress environment can be used. A couple of alternatives that you can consider, that do not require installing Docker.

Reviewer: ```suggestion
Docker is just one method to run a local WordPress environment. Block development and extending WordPress is done using normal plugins, so any WordPress environment can be used. Here are some alternatives that you can consider which do not require installing Docker.
```

## Uninstall - Start Over

Here are a few instructions if you need to start over, or want to remove what was installed.

### Local Environment

* You can remove the local environment data for a specific project:

```
wp-env destroy
```

* To completely uninstall wp-env tool:

```
npm -g uninstall @wordpress/env
```

* To uninstall Docker, or Visual Studio Code use your OS method to remove packages. For example, on Windows run "Add or remove programs". You can additionally uninstall from the Docker Desktop app, click the bug icon at the top to switch to this Troubleshoot screen. Click Uninstall or remove.

![Docker Troubleshoot Screenshot](https://developer.wordpress.org/files/2020/08/docker-uninstall-screen.png)

### Uninstall Node/NVM

To uninstall Node/NVM, delete the NVM directory, this is typically installed at `$HOME/.nvm`, delete using

```
rm -rf "$HOME/.nvm"
```

If this does not work and the `$NVM_DIR` environment variable is set you can remove using `rm -rf "$NVM_DIR"`

To clean up any installed JavaScript packages remove the global `.npm` directory at `$HOME/.npm`,

```
rm -rf "$HOME/.npm"

> ```

Reviewer 2: Great!

Btw
It would be nice to have a check that the various have been removed. Such as checking the version of the npm, node and nvm and when a not found shows up then one knows it is removed. 

That means kinda like this step which I mentioned above. I just think it would be nice to have a check to see that one has done it correctly by checking the version of each.

Quit and restart terminal.

Then check the following:

npm -v
node -v
nvm - v
v = version information.

The result when checking for version information should be a zsh: command not found.
npm, node and nvm have all been removed.

![Checking-to-see-that-npm-node-nvm-is-uninstalled](https://user-images.githubusercontent.com/5323259/92230670-d050dc80-eeab-11ea-87bc-7c6a1aa4ccb9.png)

The above gives a confirmation that one had uninstalled correctly.

Author: Updated.

Alternative editors include [Sublime Text](https://www.sublimetext.com/) that is also available across platforms, though is a commercial product; or other free alternatives include [Vim](https://www.vim.org/), [Atom](https://atom.io/), and [Notepad++](https://notepad-plus-plus.org/) all support standard JavaScript style development.

You can use any editor you're comfortable with, it is more a personal preference. The development setup for WordPress block editor is a common JavaScript environment and most editors have plugins and suppport. The key is having a way to open, edit, and save text files.

## Uninstall - Start Over

Here are a few instructions if you need to start over, or want to remove what was installed.

> ### Local Environment

Reviewer: I might add something like this along with the destroy command 

* You can clear the WordPress datbase for a specific wp-env environment:

wp-env clean all

Author: Added :+1:

> * You can remove the local environment data for a specific project:

Reviewer: ```suggestion
* You can remove the local environment completely for a specific project:
```

(trying to clarify that it's more than just "data" (i.e. more than just the database)

## Uninstall - Start Over

Here are a few instructions if you need to start over, or want to remove what was installed.

### Local Environment

* You can remove the local environment data for a specific project:

```
wp-env destroy
```

* To completely uninstall wp-env tool:

```
npm -g uninstall @wordpress/env
```

* To uninstall Docker, or Visual Studio Code use your OS method to remove packages. For example, on Windows run "Add or remove programs". You can additionally uninstall from the Docker Desktop app, click the bug icon at the top to switch to this Troubleshoot screen. Click Uninstall or remove.

![Docker Troubleshoot Screenshot](https://developer.wordpress.org/files/2020/08/docker-uninstall-screen.png)

### Uninstall Node/NVM

To uninstall Node/NVM, delete the NVM directory, this is typically installed at `$HOME/.nvm`. If the `$NVM_DIR` environment variable is set you can remove using:

```
rm -rf "$NVM_DIR"
```

or, delete directly using:

```
rm -rf "$HOME/.nvm"

> ```

Reviewer 2: I found it confusing to have two methods to uninstall nvm.

"To uninstall Node/NVM, delete the NVM directory, this is typically installed at `$HOME/.nvm`. If the `$NVM_DIR` environment variable is set you can remove using:"

`rm -rf "$NVM_DIR"`

or, delete directly using:

`rm -rf "$HOME/.nvm"`

------

Rephrasing.
"To uninstall Node/NVM, delete the NVM directory, this is typically installed at `$HOME/.nvm`.

`rm -rf "$HOME/.nvm"`

If the above does not remove nvm then you can instead use `rm -rf "$NVM_DIR"` as the environment variables were set. 

-----------

It is important to focus on one method and have a clear if that does not work then here is an option.

Author: :+1: Updated

You can also clean up any installed JavaScript packages at `$HOME/.npm`

```
rm -rf "$HOME/.npm"
```

You can also remove the `.wp-env.json` file if created.

> 

Reviewer 2: I think we can remove this line:

"You can also remove the `.wp-env.json` file if created."

As re moved the requirement from the quick install instructions.

Author: Removed.

Reviewer 2: Technically that is not correct.

"....you can confirm the uninstall worked if you can **not** run `npm -v`, `node -v`, and `nvm -v`. You should see a "command not found" error in the terminal."

Rephrasing:
"....you can confirm the uninstall worked if you run `npm -v`, `node -v`, and `nvm -v`, and see a "command not found" error in the terminal.

Here are a few instructions if you need to start over, or want to remove what was installed.

### Local Environment

* If you just want to reset and clean the WordPress database:

```
wp-env clean all
```

* To remove the local environment completely for a specific project:

```
wp-env destroy
```

* To completely uninstall wp-env tool:

```
npm -g uninstall @wordpress/env
```

* To uninstall Docker, or Visual Studio Code use your OS method to remove packages. For example, on Windows run "Add or remove programs". You can additionally uninstall from the Docker Desktop app, click the bug icon at the top to switch to this Troubleshoot screen. Click Uninstall or remove.

![Docker Troubleshoot Screenshot](https://developer.wordpress.org/files/2020/08/docker-uninstall-screen.png)

### Uninstall Node/NVM

To uninstall Node/NVM, delete the NVM directory, this is typically installed at `$HOME/.nvm`, delete using

```
rm -rf "$HOME/.nvm"
```

If this does not work and the `$NVM_DIR` environment variable is set you can remove using `rm -rf "$NVM_DIR"`

To clean up any installed JavaScript packages remove the global `.npm` directory at `$HOME/.npm`,

```
rm -rf "$HOME/.npm"
```

> Just as you confirmed the installation worked, you can confirm the uninstall worked if you try to run `npm -v`, `node -v`, and `nvm -v` you should then see a "command not found" error in the terminal.

Reviewer 2: Adding in that one needs to quit and restart terminal.

Rephrasing:
Just as you confirmed the installation worked, you can confirm the uninstall worked. First quit and restart terminal and then try to run `npm -v`, `node -v`, and `nvm -v` you should then see a "command not found" error in the terminal.

Alternative editors include [Sublime Text](https://www.sublimetext.com/) that is also available across platforms, though is a commercial product; or other free alternatives include [Vim](https://www.vim.org/), [Atom](https://atom.io/), and [Notepad++](https://notepad-plus-plus.org/) all support standard JavaScript style development.

You can use any editor you're comfortable with, it is more a personal preference. The development setup for WordPress block editor is a common JavaScript environment and most editors have plugins and suppport. The key is having a way to open, edit, and save text files.

## Uninstall - Start Over

Here are a few instructions if you need to start over, or want to remove what was installed.

### Local Environment

* You can remove the local environment data for a specific project:

```
wp-env destroy
```

> 

Reviewer 2: How does this work? Do one need to define which specific project to remove?
It would be helpful with additional instructions here.

This is what I am getting:
```
paaljoachimromdahl@Paals-MacBook-Pro ~ % wp-env destroy
✖ No .wp-env.json file found at '/Users/paaljoachimromdahl/.wp-env.json' and could not determine if '/Users/paaljoachimromdahl' is a WordPress installation, a plugin, or a theme.
paaljoachimromdahl@Paals-MacBook-Pro ~ % 
```

I did this:
Went to the gutenberg directory/folder in the root folder for the user.
Added wp-env destroy.
It gave me a message that it could not connect to Docker. (Because I did not have it running.)
I turned on Docker and ran wp-env destroy again.
It removed a WordPress environment. (That is then probably the javascript environment and has nothing to do with the gutenberg folder I have inside the root user folder.)
```
paaljoachimromdahl@Paals-MacBook-Pro ~ % cd gutenberg
paaljoachimromdahl@Paals-MacBook-Pro gutenberg % wp-env destroy
ℹ WARNING! This will remove Docker containers, volumes, and networks associated with the WordPress instance.
? Are you sure you want to continue? Yes
✖ Error while running docker-compose command.
Couldn't connect to Docker daemon. You might need to start Docker for Mac.
paaljoachimromdahl@Paals-MacBook-Pro gutenberg % wp-env destroy
ℹ WARNING! This will remove Docker containers, volumes, and networks associated with the WordPress instance.
? Are you sure you want to continue? Yes
✔ Removed WordPress environment. (in 6s 927ms)
paaljoachimromdahl@Paals-MacBook-Pro gutenberg % 
```

Bottom line is: 
To remove a specific project one needs to go to the directory such as an example `cd gutenberg` enters the gutenberg directory inside the root user folder (or no?). From there remove the WordPress environment.

Author: Did you clarify where the Gutenberg folder came from? The wp-env script does not create a gutenberg folder.

Reviewer 2: It looks I have probably forgotten to remove the Gutenberg folder from an earlier install.

I will uninstall and start over.

* To completely uninstall wp-env tool:

```
npm -g uninstall @wordpress/env
```

> 

Reviewer 2: Testing:

Btw I got a patch notice in terminal.

```
paaljoachimromdahl@Paals-MacBook-Pro gutenberg % npm -g uninstall @wordpress/env
removed 274 packages in 2.881s
```
   ╭────────────────────────────────────────────────────────────────╮
   │                                                                │
   │      New patch version of npm available! 6.14.6 → 6.14.8       │
   │   Changelog: https://github.com/npm/cli/releases/tag/v6.14.8   │
   │               Run npm install -g npm to update!                │
   │                                                                │
   ╰────────────────────────────────────────────────────────────────╯

Author: Unrelated.

* To uninstall Docker, or Visual Studio Code use your OS method to remove packages. For example, on Windows run "Add or remove programs". You can additionally uninstall from the Docker Desktop app, click the bug icon at the top to switch to this Troubleshoot screen. Click Uninstallt or remove.

> 

Reviewer 2: Uninstallt = Uninstall

Reviewer 2: Testing:

Btw I got a patch notice in terminal.

```
paaljoachimromdahl@Paals-MacBook-Pro gutenberg % npm -g uninstall @wordpress/env
removed 274 packages in 2.881s
```
   ╭────────────────────────────────────────────────────────────────╮
   │                                                                │
   │      New patch version of npm available! 6.14.6 → 6.14.8       │
   │   Changelog: https://github.com/npm/cli/releases/tag/v6.14.8   │
   │               Run npm install -g npm to update!                │
   │                                                                │
   ╰────────────────────────────────────────────────────────────────╯

Author: Fixed :+1:

![Docker Troubleshoot Screenshot](https://developer.wordpress.org/files/2020/08/docker-uninstall-screen.png)

### Uninstall Node/NVM

To uninstall Node/NVM, delete the NVM directory, this is typically installed at `$HOME/.nvm`. If the `$NVM_DIR` environment variable is set you can remove using:

```
rm -rf "$NVM_DIR"
```

> 

Reviewer 2: Result:
```
paaljoachimromdahl@Paals-MacBook-Pro gutenberg % rm -rf "$NVM_DIR"
paaljoachimromdahl@Paals-MacBook-Pro gutenberg % 
```

Reviewer 2: Btw I came across an approach that seems even better then using `rm -rf "$NVM_DIR"`
https://github.com/nvm-sh/nvm/issues/298#issuecomment-30654686

Comment:
And it is probably better to use the $NVM_DIR variable, as not everyone has nvm installed in ~/.nvm.
Oneliner:
`rm -rf $NVM_DIR ~/.npm ~/.bower`

Author: Bower is not used. I added the additional instructions for removing `.npm`

Reviewer 2: Added comments right below.

or, delete directly using:

```
rm -rf "$HOME/.nvm"
```

> 

Reviewer 2: To check that it has been deleted add the following command:
```
npm -v
node -v
```
v = version information.

Reviewer 2: I also did a `nvm -v`
and this is the result:

paaljoachimromdahl@Paals-MacBook-Pro gutenberg % nvm -v

```
Node Version Manager (v0.35.3)

Note: <version> refers to any version-like string nvm understands. This includes:
  - full or partial version numbers, starting with an optional "v" (0.10, v0.1.2, v1)
  - default (built-in) aliases: node, stable, unstable, iojs, system
  - custom aliases you define with `nvm alias foo`

 Any options that produce colorized output should respect the `--no-colors` option.

Usage:
  nvm --help                                Show this message
  nvm --version                             Print out the installed version of nvm
  nvm install [-s] <version>                Download and install a <version>, [-s] from source. Uses .nvmrc if available
    --reinstall-packages-from=<version>     When installing, reinstall packages installed in <node|iojs|node version number>
    --lts                                   When installing, only select from LTS (long-term support) versions
    --lts=<LTS name>                        When installing, only select from versions for a specific LTS line
    --skip-default-packages                 When installing, skip the default-packages file if it exists
    --latest-npm                            After installing, attempt to upgrade to the latest working npm on the given node version
    --no-progress                           Disable the progress bar on any downloads
  nvm uninstall <version>                   Uninstall a version
  nvm uninstall --lts                       Uninstall using automatic LTS (long-term support) alias `lts/*`, if available.
  nvm uninstall --lts=<LTS name>            Uninstall using automatic alias for provided LTS line, if available.
  nvm use [--silent] <version>              Modify PATH to use <version>. Uses .nvmrc if available
    --lts                                   Uses automatic LTS (long-term support) alias `lts/*`, if available.
    --lts=<LTS name>                        Uses automatic alias for provided LTS line, if available.
  nvm exec [--silent] <version> [<command>] Run <command> on <version>. Uses .nvmrc if available
    --lts                                   Uses automatic LTS (long-term support) alias `lts/*`, if available.
    --lts=<LTS name>                        Uses automatic alias for provided LTS line, if available.
  nvm run [--silent] <version> [<args>]     Run `node` on <version> with <args> as arguments. Uses .nvmrc if available
    --lts                                   Uses automatic LTS (long-term support) alias `lts/*`, if available.
    --lts=<LTS name>                        Uses automatic alias for provided LTS line, if available.
  nvm current                               Display currently activated version of Node
  nvm ls [<version>]                        List installed versions, matching a given <version> if provided
    --no-colors                             Suppress colored output
    --no-alias                              Suppress `nvm alias` output
  nvm ls-remote [<version>]                 List remote versions available for install, matching a given <version> if provided
    --lts                                   When listing, only show LTS (long-term support) versions
    --lts=<LTS name>                        When listing, only show versions for a specific LTS line
    --no-colors                             Suppress colored output
  nvm version <version>                     Resolve the given description to a single local version
  nvm version-remote <version>              Resolve the given description to a single remote version
    --lts                                   When listing, only select from LTS (long-term support) versions
    --lts=<LTS name>                        When listing, only select from versions for a specific LTS line
  nvm deactivate                            Undo effects of `nvm` on current shell
  nvm alias [<pattern>]                     Show all aliases beginning with <pattern>
    --no-colors                             Suppress colored output
  nvm alias <name> <version>                Set an alias named <name> pointing to <version>
  nvm unalias <name>                        Deletes the alias named <name>
  nvm install-latest-npm                    Attempt to upgrade to the latest working `npm` on the current node version
  nvm reinstall-packages <version>          Reinstall global `npm` packages contained in <version> to current version
  nvm unload                                Unload `nvm` from shell
  nvm which [current | <version>]           Display path to installed node version. Uses .nvmrc if available
  nvm cache dir                             Display path to the cache directory for nvm
  nvm cache clear                           Empty cache directory for nvm

Example:
  nvm install 8.0.0                     Install a specific version number
  nvm use 8.0                           Use the latest available 8.0.x release
  nvm run 6.10.3 app.js                 Run app.js using node 6.10.3
  nvm exec 4.8.3 node app.js            Run `node app.js` with the PATH pointing to node 4.8.3
  nvm alias default 8.1.0               Set default node version on a shell
  nvm alias default node                Always default to the latest available node version on a shell

Note:
  to remove, delete, or uninstall nvm - just remove the `$NVM_DIR` folder (usually `~/.nvm`)

paaljoachimromdahl@Paals-MacBook-Pro gutenberg % 

```

Reviewer 2: One more thing.
We also need to delete the .wp-env.json file added in the install step.

Author: Added removing the .wp-env.json file

Reviewer 2: Great!

Btw
It would be nice to have a check that the various have been removed. Such as checking the version of the npm, node and nvm and when a not found shows up then one knows it is removed. 

That means kinda like this step which I mentioned above. I just think it would be nice to have a check to see that one has done it correctly by checking the version of each.

Quit and restart terminal.

Then check the following:

npm -v
node -v
nvm - v
v = version information.

The result when checking for version information should be a zsh: command not found.
npm, node and nvm have all been removed.
